### ISSUE TITLE
[red-knot] add Type::SubclassOf support to Type::call and Type::call_unbound

### ISSUE THREAD

**Issue by carljm** [astral-sh/ruff#15948]
### Description

Existing type checkers (and the typing spec) allow calling an object of type `type[C]` (in red-knot, `Type::SubclassOf(<Class C>)`), with the result of the call being an object of type `C` (in red-knot, `Type::Instance(<Class C>)`).

This is not actually sound, because LSP is not enforced on type constructor methods (`__init__` and `__new__`), so you can't be sure that a constructor call that works for a class `C` will actually work for a subclass of `C`. But it's widely relied on, and we will have to support it.

I would like to have an opt-in diagnostic emitted whenever you call an object of `type[C]`, so users who want to avoid this unsoundness have that option.

**Comment by carljm** [comment]
On second thought, it may be best to hold off on this issue (or at least the new-diagnostic part of it) until after @AlexWaygood completes an in-progress refactor of `CallOutcome`, since this diagnostic would likely imply adding a new outcome variant to `CallOutcome`?

**Comment by mishamsk** [comment]
I can take it whenever @AlexWaygood is ready since I've triggered the issue in the first place. Not sure how you manage blocked issues, but feel free to assign this one to me and I'll monitor the comments

**Comment by mishamsk** [comment]
@AlexWaygood I believe this one is unblocked now? Also see another comment related to analyzing `__init__` signatures for class literals:

```
// TODO annotated return type on `__new__` or metaclass `__call__`
            // TODO check call vs signatures of `__new__` and/or `__init__`
```

which may or may not be tackled together...

**Comment by AlexWaygood** [comment]
> [@AlexWaygood](https://github.com/AlexWaygood) I believe this one is unblocked now?

Should be, yup!

Those other todos should be tackled in separate PRs, I think; they're pretty separate issues

**Comment by sharkdp** [comment]
The core functionality was added in #16597, but the diagnostic has not been implemented. Also, those calls will really only be properly checked when #16511 is also implemented.

**Comment by mishamsk** [comment]
I think the full implementation will be pretty involved, but may be out of scope here. I was going to ask related question this week.

What I was thinking about is `Self`. When we handle plain class literals (#16511 case) we can avoid properly handling `Self` since `class.to_instance()` will return the correct type in all realistic scenarios. But with `type[C]`, in ideal world we would record a stronger typing relationship - due to the free type variable... so fully model this:

```py
T = TypeVar("T", bound="Base")

class Base:
    pass

class Sub(Base):
    pass

def get_instance(clz: type[T]) -> T:
    return clz()

instance = get_instance(Sub)
reveal_type(instance) # Sub, not Base
```

but for constructors.

and AFAIK there is no machinery for that currently. 

Anyways, that's for the future. I hope I'll close #16511 this week, and we can decide whether it makes sense to do a small PR to add the diagnostic here or dive into `Self` support

**Comment by mishamsk** [comment]
yeah, I thought so. Is there an issue I could track to see the progress related to type variables / generics?

**Comment by AlexWaygood** [comment]
> yeah, I thought so. Is there an issue I could track to see the progress related to type variables / generics?

(Sent you an invite to our internal design doc around this!)

**Comment by mishamsk** [comment]
thanks a lot! fascinating read. I'll keep an eye on the roadmap, maybe I'll find a non-impending way to positively slot in a PR or smth. somewhere along the way!

**Comment by AlexWaygood** [comment]
> > I would like to have an opt-in diagnostic emitted whenever you call an object of type[C], so users who want to avoid this unsoundness have that option.
> 
> I think this comment by [@carljm](https://github.com/carljm?rgh-link-date=2025-03-31T11%3A26%3A03.000Z) is the only part of this issue that is not yet covered by things already implemented or things tracked by other issues. I'll open a followup issue to track this specifically.

I wroteup https://github.com/astral-sh/ruff/issues/17089 for this feature
